### PR TITLE
Add 'welcome' app for first boot

### PR DESCRIPTION
**Pull request by jcapona**
| Status  | Ticket/Issue |
| :---: | :--: |
| Ready/Hold | [Ticket](https://pi-top.atlassian.net/browse/SWE-400) |

#### Main changes
- Add 'welcome' app that runs only on first boot.
- Prompt users to go to pi-top.com/start.
- Only runs on devices running pi-topOS.

Associated PR: https://github.com/pi-top/pi-top-Python-SDK/pull/695

#### Screenshots (feature, test output, profiling, dev tools etc)
first_page

second_page

#### Other notes (e.g. implementation quirks, edge cases, questions / issues)
-

#### Manual testing tips
-

#### Tag anyone who definitely needs to review or help
-

### CODE REVIEW

File: debian/pt-miniscreen.postinst [hunk 1]
```diff
@@ -11,6 +11,20 @@ case "${1}" in
   configure)
     LAST_UPDATE_FOLDER="/var/lib/pt-miniscreen"
     LAST_UPDATE_FILE="${LAST_UPDATE_FOLDER}/last-update"
+    STATE_FILE="${LAST_UPDATE_FOLDER}/state.cfg"
+
+    if [ ! -z "$2" ]; then
+	  # Upgrading from a previous version...
+      if [ ! -f "${STATE_FILE}" ]; then
+        # Creating state file to avoid running the welcome app
+		mkdir -p "${LAST_UPDATE_FOLDER}"
```
Comment (stefan-kreitmayer) [inline debian/pt-miniscreen.postinst:20]: pls fix indentation, thx

File: tests/test_state.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+
+def test_singleton_pattern():
+    manager1 = StateManager("test_package", testing=True)
+    manager2 = StateManager("test_package", testing=True)
+    assert manager1 is manager2
+
+
+def test_state_file_creation(state_manager):
+    state_file = Path("/tmp/test_package/state.cfg")
```
Comment (stefan-kreitmayer) [inline tests/test_state.py:30]: string used in the path should be based on random uuid generated in the state manager. so we should be checking for that path

File: tests/test_state.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+def test_set_creates_section(state_manager):
+    # Test that set() creates a new section if it doesn't exist
+    state_manager.set("new_section", "key", "value")
+    assert state_manager.get("new_section", "key") == "value"
+
+
+def test_set_raises_on_invalid_input(state_manager):
+    # Test that set() raises an exception with invalid input
+    with pytest.raises(Exception):
+        state_manager.set(None, "key", "value")
```
Comment (stefan-kreitmayer) [inline tests/test_state.py:93]: consider adding more invalid inputs
- empty string (should raise)
- number (should be parsed as string)
- bytes (should raise)
- boolean (should raise)

Comment (stefan-kreitmayer) [reply tests/test_state.py:93]: essentially we expect to raise whenever the config parser raises

File: pt_miniscreen/state.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        with self.lock:
+            val = fallback
+            try:
+                val = self.config_parser.get(section, key)
+            except Exception:
+                if fallback is None:
+                    raise
+            return val
+
+    def set(self, section: str, key: str, value):
```
Comment (stefan-kreitmayer) [inline pt_miniscreen/state.py:54]: optionally: consider checking for invalid input types 
(maybe look into config parser to see how that is expected to behave)

File: pt_miniscreen/state.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+                raise
+            self._save()
+
+    def _save(self):
+        with open(self.state_file_path, "w") as f:
+            self.config_parser.write(f)
+
+    @classmethod
+    def exists(cls, package_name: str, testing: Optional[bool] = None):
+        state_file_dir = "/tmp" if testing else "/var/lib"
```
Comment (stefan-kreitmayer) [inline pt_miniscreen/state.py:81]: consider extracting `"/tmp" if testing else "/var/lib"` into a method to avoid duplication

File: pt_miniscreen/state.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
+import logging
```
Comment (stefan-kreitmayer) [inline pt_miniscreen/state.py:1]: consider moving this class into the SDK

Comment (stefan-kreitmayer) [reply pt_miniscreen/state.py:1]: and then create a ticket for web portal to use the SDK version of this class

Comment (stefan-kreitmayer) [reply pt_miniscreen/state.py:1]: maybe do both in a separate (cross repo) refactor ticket

File: tests/test_state.py (hunk 3 above)
Comment (stefan-kreitmayer) [reply tests/test_state.py:93]: let's raise on 
- empty string
- anything that isn't a string, including numbers

File: pt_miniscreen/state.py [hunk 7]
```diff
@@ -1,45 +1,66 @@
...
-            cls._instance = super().__new__(cls)
-        # python will automatically call __init__ after __new__
-        return cls._instance
-
-    def __init__(self, package_name: str, testing: Optional[bool] = None):
-        # Skip initialization if already done
-        if StateManager._initialized:
-            return
+    _instances: Dict[str, "StateManager"] = {}
+    _initialized: Set[str] = set()
```
Comment (stefan-kreitmayer) [inline pt_miniscreen/state.py]: consider replacing the class-scope set with a instance variable (boolean)

File: pt_miniscreen/state.py [hunk 8]
```diff
@@ -1,45 +1,66 @@
...
+            cls._instances[name] = super().__new__(cls)
+        return cls._instances[name]
 
-        state_file_dir = "/tmp" if testing else "/var/lib"
-        self.state_file_path = f"{state_file_dir}/{package_name}/state.cfg"
-        self.config_parser = ConfigParser()
-        self.lock = Lock()
+    def __init__(self, name: str):
+        # Skip initialization if already done for this package
+        if name in StateManager._initialized:
```
Comment (stefan-kreitmayer) [inline pt_miniscreen/state.py]: this will be something like `if self._initialized:`

File: pt_miniscreen/state.py [hunk 9]
```diff
@@ -72,12 +97,22 @@ def remove(self, section: str, key: str = ""):
...
+    def remove_section(self, section: str):
+        if section not in self.config_parser.sections():
+            return
+        with self.lock:
+            self.config_parser.remove_section(section)
+            self._save()
+
+    def remove_key(self, section: str, key: str):
+        if section not in self.config_parser.sections():
+            return
```
Comment (stefan-kreitmayer) [inline pt_miniscreen/state.py]: consider raising if section or key doesn't exist

File: tests/test_state.py [hunk 10]
```diff
@@ -7,77 +7,79 @@
...
     if state_file.exists():
         state_file.unlink()
     if state_file.parent.exists():
         state_file.parent.rmdir()
 
 
 def test_singleton_pattern():
-    manager1 = StateManager("test_package", testing=True)
-    manager2 = StateManager("test_package", testing=True)
+    # Test that the state manager is a singleton
```
Comment (stefan-kreitmayer) [inline tests/test_state.py]: maybe rephrase "Test that names refer to the same instance of state manager"

File: tests/test_state.py [hunk 11]
```diff
@@ -7,77 +7,79 @@
...
+    manager3 = StateManager("some_other_name")
+    assert manager1 is not manager3
+
+    # cleanup
+    state_file = Path(manager1.path)
+    if state_file.exists():
+        state_file.unlink()
+    if state_file.parent.exists():
+        state_file.parent.rmdir()
+
```
Comment (stefan-kreitmayer) [inline tests/test_state.py]: cleanup manager 3 as well

File: tests/test_state.py [hunk 12]
```diff
@@ -89,15 +91,20 @@ def test_set_creates_section(state_manager):
...
+        True,
+        object(),
+        [1, 2],
+        {"hey": "ho"},
+        "",
+        b"somedata",
+    ]:
+        with pytest.raises(ValueError):
+            state_manager.set("some_section", data, "value")
+            state_manager.set(data, "some_key", "value")
```
Comment (stefan-kreitmayer) [inline tests/test_state.py]: split up in two separate tests
test_remove_raises_with_invalid_section
test_remove_raises_with_invalid_key

File: tests/test_state.py [hunk 13]
```diff
@@ -106,3 +113,43 @@ def test_multiple_values_in_section(state_manager):
     state_manager.set("test_section", "key2", "value2")
     assert state_manager.get("test_section", "key1") == "value1"
     assert state_manager.get("test_section", "key2") == "value2"
+
+
+def test_remove_key_doesnt_raise_on_nonexistent_key(state_manager):
+    # Test that remove_key() doesnt raise an exception on nonexistent key
+    state_manager.remove_key("nonexistent_section", "nonexistent_key")
```
Comment (stefan-kreitmayer) [inline tests/test_state.py]: change accordingly, see other comment

File: tests/test_state.py [hunk 14]
```diff
@@ -106,3 +113,43 @@ def test_multiple_values_in_section(state_manager):
...
+
+
+def test_remove_section_doesnt_raise_on_nonexistent_section(state_manager):
+    # Test that remove_section() doesnt raise an exception on nonexistent section
+    state_manager.remove_section("nonexistent_section")
+
+
+def test_remove_section_remvoes_all_keys(state_manager):
```
Comment (stefan-kreitmayer) [inline tests/test_state.py]: typo :)

File: tests/test_state.py [hunk 15]
```diff
@@ -106,3 +113,43 @@ def test_multiple_values_in_section(state_manager):
...
+    # Check that both keys are removed
+    with pytest.raises(Exception):
+        state_manager.get("test_section", "key1")
+    with pytest.raises(Exception):
+        state_manager.get("test_section", "key2")
+
+
+def test_remove_key_removes_key_from_section(state_manager):
+    # Test that remove_key() removes a key
+    state_manager.set("test_section", "key1", "value1")
```
Comment (stefan-kreitmayer) [inline tests/test_state.py]: add key2 
verify that key1 was removed
verify that key2 was not removed
